### PR TITLE
post: typescript part 2

### PR DESCRIPTION
**Pull request by cexbrayat**
This is the post for next tuesday: TS part 2.

I rewrote the decorator part and added a few things.
If you find it OK, I'll push a PR for the book with the same stuff.

### CODE REVIEW

File: _posts/2015-11-17-typescript-part-2.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+This is the second part in [our discovery of TypeScript](/2015/11/12/typescript-part-1/).
+As we saw, TypeScript is a superset of ES6, so it implements classes,
+and adds a few bonuses on the way.
+
+# Classes
+
+A class can implement an interface. For us, the `Pony` class should be able to run, so we can do:
+
+    class Pony implements CanRun {
```
Comment (clacote) [inline _posts/2015-11-17-typescript-part-2.md]: It's strange to use an interface defined in a previous post. But... that's too late, and not so bad.

File: _posts/2015-11-17-typescript-part-2.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    // defines a public property name with 'Rainbow Dash'
+    // and a private one speed with 10    
+
+These shortcuts are really useful and we'll rely on them a lot in Angular 2!
+
+# Working with other libraries
+
+When working with external libraries written in JS, you may think we are doomed because we don't know what types of parameter the function in that library will expect. That's one of the cool things with the TypeScript community: its members have defined interfaces for the types and functions exposed by the popular JavaScript libraries!
+
+The files containing these interfaces have a special `.d.ts` extension. They contain a list of the library's public functions. A good place to look for these files is https://github.com/borisyankov/DefinitelyTyped[DefinitelyTyped]. For example, if you want to use TS in your AngularJS 1.x apps, you can download the proper file from the repo:
```
Comment (clacote) [inline _posts/2015-11-17-typescript-part-2.md]: Bad link syntax: 
image

File: _posts/2015-11-17-typescript-part-2.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+Generally, annotations are not really used by the language itself, but mainly by frameworks and libraries.
+
+Decorators are really powerful: they can modify their target, and for example to add metadata.
+Until now, it was not something possible in JavaScript.
+But the language is evolving and there is now an official proposal for `decorators`,
+that may be standardized one day in the future (possibly in ES7/ES2016).
+Note that the TypeScript implementation goes slightly further than the proposed standard.
+
+In Angular 2, we will use the decorators provided by the framework.
+They role is fairly basic: they add some metadata to our classes to say for example "this class is a component",
```
Comment (clacote) [inline _posts/2015-11-17-typescript-part-2.md]: "They role" -> "Their role"

File: _posts/2015-11-17-typescript-part-2.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+    }
+
+The `@Component` decorator is added to the class `Home`.
+When Angular 2 will load our app, it will find the class `Home`,
+and will understand that it is a component, based on the metadata the decorator will add.
+Cool, huh? As you can see, a decorator can also receive parameters, here a configuration object.
+
+I just wanted to introduce the raw concept of decorators,
+we'll look into every decorator available in Angular all along the book.
```
Comment (clacote) [inline _posts/2015-11-17-typescript-part-2.md]: Which book? :p

You'll have to either link to the ebook page there, or to change this sentence to not mention a book in a blog post series.

File: _posts/2015-11-17-typescript-part-2.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+I have to point out that you can use decorators with Babel as a transpiler instead of TypeScript.
+There is even a plugin to support all the Angular 2 deocrators: [angular2-annotations](https://www.npmjs.com/package/babel-plugin-angular2-annotations)
```
Comment (clacote) [inline _posts/2015-11-17-typescript-part-2.md]: Lacks a "." at the end of the sentence.

File: _posts/2015-11-17-typescript-part-2.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+Babel also supports class properties, but not the type system offered by TypeScript.
+You can use Babel, and write "ES6+" code, but you will not be able to use the types, and they are very useful for the dependency injection for example.
+It's completely possible, but you'll have to add more decorators to replace the types.
+
+So my advice would be to give TypeScript a try!
+All my examples from here will use it.
```
Comment (clacote) [inline _posts/2015-11-17-typescript-part-2.md]: Same as ebook mention: there will be no more examples on the blog. Just remove this sentence.

File: _posts/2015-11-17-typescript-part-2.md (hunk 1 above)
Comment (cexbrayat) [inline _posts/2015-11-17-typescript-part-2.md]: Yeah I figure it out too late...

**Comment by cexbrayat** [comment]
new commit with your feedback

@jnizet maybe you want to have a look at the new part about decorators, I think you pointed out that it was not really clear before :)

File: _posts/2015-11-17-typescript-part-2.md [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+So you don't even have to worry about including the interfaces in your Angular 2 project:
+the TS compiler will figure it out by itself if you are using NPM to manage your dependencies!
+
+# Decorators
+
+This is a fairly new feature, added only in TypeScript 1.5, to help supporting Angular.
+Indeed, as we will shortly see, Angular 2 components can be described using decorators.
+You may have not heard about decorators, as not every language has them.
+A decorator is a way to do some meta-programming.
+There are fairly similar to annotations
```
Comment (jnizet) [inline _posts/2015-11-17-typescript-part-2.md]: They

File: _posts/2015-11-17-typescript-part-2.md [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+which are mainly used in Java, #C and Python, and maybe other languages I don't know.
```
Comment (jnizet) [inline _posts/2015-11-17-typescript-part-2.md]: C#

File: _posts/2015-11-17-typescript-part-2.md [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+Depending on the language, you add an annotation to a method, an attribute, or a class.
+Generally, annotations are not really used by the language itself, but mainly by frameworks and libraries.
+
+Decorators are really powerful: they can modify their target, and for example to add metadata.
```
Comment (jnizet) [inline _posts/2015-11-17-typescript-part-2.md]: What does "modify their target" mean? AFAIK, the *only* thing they can do, is adding metadata. I'd rephrase:
> Until now, it was not possible to associate metadata to a JavaScript class or function. But the language is evolving and ...

File: _posts/2015-11-17-typescript-part-2.md [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+Their role is fairly basic: they add some metadata to our classes to say for example "this class is a component",
+"this is an optional dependency", "this is a custom property", etc...
+It's not required to use them, as you can add the metadata manually (if you want to stick to ES5 for example),
+but the code will be definitely more elegant using decorators, as provided by TypeScript.
+
+In TypeScript, decorators start with an `@`, and can be applied to a class, a class property, a function or a function parameter.
+Not on a constructor though, but it can be applied to the constructor's parameters.
+
+To have a better grasp on this, let's try to build a simple decorator,
+`@log()`, that will log something every time a method is called.
```
Comment (jnizet) [inline _posts/2015-11-17-typescript-part-2.md]: Isn't it a best practice to make them start with an uppercase letter?

File: _posts/2015-11-17-typescript-part-2.md [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+It will be use like this:
```
Comment (jnizet) [inline _posts/2015-11-17-typescript-part-2.md]: used

File: _posts/2015-11-17-typescript-part-2.md [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+Depending on what you want to apply your decorator, the function will not have exactly the same arguments.
+Here we have a method decorator, that takes 3 parameters:
+
+- `target`: the method targeted by our decorator
+- `name`: the name of the targeted method
+- `descriptor`: a descriptor of the targeted method, like is the method enumerable, writable, etc...
+
+Here we simply log the method name, but you could do pretty much whatever you want:
+interfere with the parameters, the result, calling another function, etc...
```
Comment (jnizet) [inline _posts/2015-11-17-typescript-part-2.md]: I don't understand what the log function does, nor when it's called, and by whom. I also don't understand how the function and the decorator are linked together. Is the function supposed to be a global function with the same name as the decorator? Please tell me they didn't do that.

File: _posts/2015-11-17-typescript-part-2.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+Babel also supports class properties, but not the type system offered by TypeScript.
+You can use Babel, and write "ES6+" code, but you will not be able to use the types, and they are very useful for the dependency injection for example.
+It's completely possible, but you'll have to add more decorators to replace the types.
+
+So my advice would be to give TypeScript a try!
+It's not very intrusive, as you can use it just where it's useful and forget about it for the rest.
+If you really don't like it, it will not be very difficult to switch to ES6 with Babel or Traceur,
+or even ES5, if you are slightly crazy (but honestly, an Angular 2 app in ES5 has pretty ugly code).
+
+Stay tune for the next episode of 'The road to Angular 2'!
```
Comment (jnizet) [inline _posts/2015-11-17-typescript-part-2.md]: Stay tuned

File: _posts/2015-11-17-typescript-part-2.md (hunk 12 above)
Comment (cexbrayat) [inline _posts/2015-11-17-typescript-part-2.md]: Every time the decorated method is called, the decorator will do a console.log with the name of the decorated method.
The decorator is a function like the other: it can be a global one, or one in the module, or in a dedicated module imported. I'll try to clarify.

File: _posts/2015-11-17-typescript-part-2.md (hunk 10 above)
Comment (cexbrayat) [inline _posts/2015-11-17-typescript-part-2.md]: Good question. I did not read something about that, but as the decorator in Angular 2 are uppercased, I'll do the same.

File: _posts/2015-11-17-typescript-part-2.md (hunk 9 above)
Comment (cexbrayat) [inline _posts/2015-11-17-typescript-part-2.md]: They can do what they want with the parameters, or alter the result, etc... But Angular 2 uses them to add metadata.
I'll clarify.

**Comment by cexbrayat** [comment]
@jnizet tried to improve with your feedback in 0c21ff6a46bf35306382cae36f4e6875088c8b5a

File: _posts/2015-11-17-typescript-part-2.md (hunk 12 above)
Comment (jnizet) [inline _posts/2015-11-17-typescript-part-2.md]: Ah, so that explains why they're called decorators and not annotations.
They're in fact closer to AOP interceptors than to annotations.
I also understand how the link is made now: through TypeScript imports
(which I tend to forget about, because I'm so used to ES5).
Thanks for the clarification.

2015-11-15 10:44 GMT+01:00 Cédric Exbrayat <notifications@github.com>:

**Comment by cexbrayat** [comment]
@clacote merged and ready for your PR skills

**Comment by clacote** [comment]
https://twitter.com/ninjasquad/status/666511664154591232
